src/ordered_dict.jl

```diff
@@ -84,9 +83,44 @@ function OrderedDict(kv)
     end
 end
 
-dict_with_eltype{K,V}(kv, ::Type{Tuple{K,V}}) = OrderedDict{K,V}(kv)
-dict_with_eltype{K,V}(kv, ::Type{Pair{K,V}}) = OrderedDict{K,V}(kv)
-dict_with_eltype(kv, t) = OrderedDict{Any,Any}(kv)
+@static if VERSION >= v"0.5.0"
+    OrderedDict{K,V}(kv::Tuple{Vararg{Pair{K,V}}})         = OrderedDict{K,V}(kv, iteratorsize(kv))
+    OrderedDict{K}(kv::Tuple{Vararg{Pair{K}}})             = OrderedDict{K,Any}(kv, iteratorsize(kv))
+    OrderedDict{V}(kv::Tuple{Vararg{Pair{TypeVar(:K),V}}}) = OrderedDict{Any,V}(kv, iteratorsize(kv))
+    OrderedDict(kv::Tuple{Vararg{Pair}})                   = OrderedDict{Any,Any}(kv, iteratorsize(kv))
+
+    OrderedDict{K,V}(kv::AbstractArray{Tuple{K,V}}) = OrderedDict{K,V}(kv, iteratorsize(kv))
+    OrderedDict{K,V}(kv::AbstractArray{Pair{K,V}})  = OrderedDict{K,V}(kv, iteratorsize(kv))
+    OrderedDict{K,V}(kv::Associative{K,V})          = OrderedDict{K,V}(kv, iteratorsize(kv))
+
+    OrderedDict{K,V}(ps::Pair{K,V}...)          = OrderedDict{K,V}(ps, iteratorsize(ps))
+    OrderedDict{K}(ps::Pair{K}...,)             = OrderedDict{K,Any}(ps, iteratorsize(ps))
+    OrderedDict{V}(ps::Pair{TypeVar(:K),V}...,) = OrderedDict{Any,V}(ps, iteratorsize(ps))
+    OrderedDict(ps::Pair...)                    = OrderedDict{Any,Any}(ps, iteratorsize(ps))
+
+    dict_with_eltype{K,V}(kv, ::Type{Tuple{K,V}}) = OrderedDict{K,V}(kv, iteratorsize(kv))
+    dict_with_eltype{K,V}(kv, ::Type{Pair{K,V}}) = OrderedDict{K,V}(kv, iteratorsize(kv))
+    dict_with_eltype(kv, t) = OrderedDict{Any,Any}(kv, iteratorsize(kv))
```

Comment:
This block can be removed if you define `get_n_slotsz` above.  (There is probably a better name for that function...)